Fix orphan page issue in _bt_split

knizhnik opened the pull request.

HaoyuHuang commented on src/backend/access/nbtree/nbtinsert.c:
```diff
@@ -1545,8 +1548,8 @@ _bt_split(Relation rel, Relation heaprel, BTScanInsert itup_key, Buffer buf,
-	/* Allocate temp buffer for leftpage */
-	leftpage = PageGetTempPage(origpage);
+	leftpage = leftpage_buf.data;
+	memcpy(leftpage, origpage, BLCKSZ);
```
do we need to memcpy? PageGetTempPage only palloc a page.

MMeent replied:
Yes, I think that memcpy can be removed, or replaced with `memset(leftpage, 0, BLCKSZ);` if we want to make sure we don't leak values that previously populated the stack.

knizhnik replied:
Sorry, this memcpy is really not needed.
Removed.

HaoyuHuang reviewed (commented):
Thanks for the failpoints. :)

knizhnik committed:
Improve stability of btree page split on ERRORs
This improves the stability of VACUUM when processing btree indexes,
which was previously able to trigger an assertion failure in
_bt_lock_subtree_parent() when an error was previously thrown outside
the scope of _bt_split() when splitting a btree page.  VACUUM would
consider the index as in a corrupted state as the right page would not
be zeroed for the error thrown (allocation failure is one pattern).

In a non-assert build, VACUUM is able to succeed, reporting what it sees
as a corruption while attempting to fix the index.  This would manifest
as a LOG message, as of:
LOG: failed to re-find parent key in index "idx" for deletion target
page N
CONTEXT:  while vacuuming index "idx" of relation "public.tab"

This commit improves the code to rely on two PGAlignedBlocks that are
used as a temporary space for the left and right pages.  The main change
concerns the right page, whose contents are now copied into the
"temporary" PGAlignedBlock page while its original space is zeroed.  Its
contents are moved from the PGAlignedBlock page back to the page once we
enter in the critical section used for the split.  This simplifies the
split logic, as it is not necessary to zero the right page before
throwing an error anymore.  Hence errors can now be thrown outside the
split code.  For the left page, this shaves one allocation, with
PageGetTempPage() being previously used.

The previous logic originates from commit 8fa30f9, at a point where
PGAlignedBlock did not exist yet.  This could be argued as something
that should be backpatched, but the lack of complaints indicates that it
may not be necessary.

Author: Konstantin Knizhnik <knizhnik@garret.ru>
Discussion: https://postgr.es/m/566dacaf-5751-47e4-abc6-73de17a5d42a@garret.ru
